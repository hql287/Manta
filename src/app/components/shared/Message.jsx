// React Libraries
import React from 'react';
import PropTypes from 'prop-types';

// Styles
import styled from 'styled-components';
const MessageStyle = styled.div`
  padding: 20px;
  font-weight: 200;
  color: #4f555c;
  background: white;
  border-radius: 4px;
  border: 1px solid rgba(0, 0, 0, 0.1);
  ${props => props.info && `border-left: 5px solid #469FE5;`} ${props =>
      props.success && `border-left: 5px solid #6BBB69;`} ${props =>
      props.danger && `border-left: 5px solid #EC476E;`} ${props =>
      props.warning && `border-left: 5px solid #F9D548;`};
`;

// Component
const Message = props => <MessageStyle {...props}>{props.text}</MessageStyle>;

Message.propTypes = {
  text: PropTypes.string.isRequired,
  info: PropTypes.bool,
  success: PropTypes.bool,
  danger: PropTypes.bool,
  warning: PropTypes.bool,
};

Message.defaultProps = {
  info: true,
  success: false,
  danger: false,
  warning: false,
};

export default Message;
