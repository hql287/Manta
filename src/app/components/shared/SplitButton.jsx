// Libraries
import React, { Component } from 'react';
import PropTypes from 'prop-types';

// Styles
import styled from 'styled-components';

const Wrapper = styled.div`
  position: relative;
`;

const ButtonsGroup = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  > a {
    flex: 1;
    border: 1px solid rgba(0, 0, 0, 0.1);
    display: flex;
    justify-content: center;
    // align-items: center;
    padding: 4px 0;
    font-size: 12px;
    text-decoration: none;
    background: #ffffff;
    text-transform: uppercase;
    letter-spacing: 1px;
    color: #000;
    &:first-child {
      border-radius: 4px 0 0 4px;
      padding: 4px 12px;
      text-align: left;
      justify-content: flext-start;
      // align-items: flext-start;
    }
    &:last-child {
      border-radius: 0 4px 4px 0;
      border-left: none;
      max-width: 24px;
    }
  }
`;

const Addon = styled.div`
  &:before {
    content: ' ';
    height: 10px;
    width: 10px;
    position: absolute;
    top: -6px;
    right: 10px;
    transform: rotate(45deg);
    background: white;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-left: 1px solid rgba(0, 0, 0, 0.1);
  }
  display: flex;
  flex-direction: column;
  position: absolute;
  right: 0;
  top: 42px;
  width: 100%;
  background: white;
  border-radius: 4px;
  border: 1px solid rgba(0, 0, 0, 0.1);
  z-index: +999;
  a {
    text-transform: uppercase;
    font-size: 12px;
    letter-spacing: 1px;
    padding: 6px 12px;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    color: black;
    width: 100%;
    text-align: left;
    z-index: 10;
    &:hover {
      text-decoration: none;
      background: #f9fafa;
    }
    &:last-child {
      border-bottom: none;
    }
  }
`;

class SplitButton extends Component {
  constructor(props) {
    super(props);
    this.state = { showOptions: false };
    this.handleMainAction = this.handleMainAction.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.handleOutsideClick = this.handleOutsideClick.bind(this);
    this.renderListItems = this.renderListItems.bind(this);
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (this.state.showOptions !== nextState.showOptions) {
      return true;
    }
    return false;
  }

  handleClick() {
    if (!this.state.showOptions) {
      document.addEventListener('click', this.handleOutsideClick, false);
    } else {
      document.removeEventListener('click', this.handleOutsideClick, false);
    }
    this.setState({ showOptions: !this.state.showOptions });
  }

  handleOutsideClick(e) {
    if (this.node.contains(e.target)) {
      return;
    }
    this.handleClick();
  }

  handleMainAction(e) {
    this.setState({ showOptions: false }, () => {
      this.props.mainButton.action();
    });
  }

  renderListItems() {
    const { options } = this.props;
    return options.map(option => (
      <a
        key={option.label.toString()}
        href="#"
        onClick={() => {
          option.action();
          this.handleClick();
        }}
      >
        {option.label}
      </a>
    ));
  }

  render() {
    return (
      <div ref={node => (this.node = node)}>
        <Wrapper>
          <ButtonsGroup>
            <a href="#" onClick={this.handleMainAction}>
              {this.props.mainButton.label}
            </a>
            <a href="#" onClick={this.handleClick}>
              <i className="ion-arrow-down-b" />
            </a>
          </ButtonsGroup>
          {this.state.showOptions && <Addon>{this.renderListItems()}</Addon>}
        </Wrapper>
      </div>
    );
  }
}

SplitButton.propTypes = {
  mainButton: PropTypes.shape({
    label: PropTypes.string.isRequired,
    action: PropTypes.func.isRequired,
  }).isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      action: PropTypes.func.isRequired,
    })
  ).isRequired,
};

export default SplitButton;
