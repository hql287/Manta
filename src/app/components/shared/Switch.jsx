// Libraries
import React from 'react';
import PropTypes from 'prop-types';

// Styles
import styled from 'styled-components';

const SwitchStyle = styled.label`
  position: relative;
  display: inline-block;
  width: 46px;
  height: 24px;
  margin: 0;
`;

const Input = styled.input`
  display: none;
  &:checked + span {
    background-color: #6bbb69;
  }
  &:checked + span:before {
    -webkit-transform: translateX(20px);
    -ms-transform: translateX(20px);
    transform: translateX(20px);
  }
`;

const Slider = styled.span`
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: #c4c8cc;
  -webkit-transition: 0.4s;
  transition: 0.4s;
  &::before {
    position: absolute;
    content: '';
    height: 18px;
    width: 18px;
    left: 4px;
    bottom: 3px;
    background-color: white;
    -webkit-transition: 0.4s;
    transition: 0.4s;
  }
  ${props =>
    props.rounded &&
    `
    border-radius: 34px;
    &::before {
      border-radius: 50%;
    }
  `};
`;

export const Switch = props => (
  <SwitchStyle>
    <Input type="checkbox" {...props} />
    <Slider rounded />
  </SwitchStyle>
);

Switch.propTypes = {
  checked: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
};

export default Switch;
