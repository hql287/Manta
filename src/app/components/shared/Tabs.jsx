// Libraries
import React from 'react';

// Styles
import styled from 'styled-components';

const TabsStyle = styled.div`
  display: flex;
  justify-content: flex-start;
`;

const TabStyle = styled.a`
  background: #f9fbfa;
  text-decoration: none;
  color: #292b2c;
  font-weight: 400;
  font-size: 16px;
  border-right: 1px solid rgba(0, 0, 0, 0.1);
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  padding: 6px 12px;
  &:last-child {
    border-right: 1px solid rgba(0, 0, 0, 0.1);
  }
  &.active {
    background: #fff;
    border-bottom: 1px solid #fff;
  }
  &:hover {
    text-decoration: none;
    color: #292b2c;
  }
`;

const TabContentStyle = styled.div`
  padding: 40px;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  margin-top: -1px;
  border-top-left-radius: 0;
`;

// Components
const Tab = props => <TabStyle {...props}>{props.children}</TabStyle>;

const TabContent = props => <TabContentStyle>{props.children}</TabContentStyle>;

const Tabs = props => <TabsStyle>{props.children}</TabsStyle>;

export { Tab, Tabs, TabContent };
