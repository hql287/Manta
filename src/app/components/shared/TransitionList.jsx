// Libs
import React, { Component } from 'react';
import PropTypes from 'prop-types';

// React Motion
import { TransitionMotion, spring } from 'react-motion';

// HOC Componen
class TransitionList extends Component {
  constructor(props) {
    super(props);
    this.getDefaultStyles = this.getDefaultStyles.bind(this);
    this.getStyles = this.getStyles.bind(this);
    this.willEnter = this.willEnter.bind(this);
    this.willLeave = this.willLeave.bind(this);
  }

  getDefaultStyles() {
    const { children, componentHeight } = this.props;
    const defaultStyles = children.map(child => ({
      key: child.key,
      data: { child },
      style: {
        height: componentHeight,
        opacity: 0,
      },
    }));
    return defaultStyles;
  }

  getStyles() {
    const { children, componentHeight } = this.props;
    const styles = children.map(child => ({
      key: child.key,
      data: { child },
      style: {
        height: spring(componentHeight),
        opacity: spring(1),
      },
    }));
    return styles;
  }

  willEnter() {
    return {
      height: 0,
      opacity: 0,
    };
  }

  willLeave() {
    return {
      height: spring(0),
      opacity: spring(0),
    };
  }

  render() {
    return (
      <TransitionMotion
        willEnter={this.willEnter}
        willLeave={this.willLeave}
        styles={this.getStyles()}
        defaultStyles={this.getDefaultStyles()}
      >
        {values => (
          <div>
            {values.map(({ key, style, data }) => (
              <div key={key} style={style}>
                {data.child}
              </div>
            ))}
          </div>
        )}
      </TransitionMotion>
    );
  }
}

TransitionList.propTypes = {
  children: PropTypes.arrayOf(PropTypes.element).isRequired,
  componentHeight: PropTypes.number.isRequired,
};

export default TransitionList;
