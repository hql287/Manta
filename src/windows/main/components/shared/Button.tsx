// Libraries
import React, { FC, ReactNode } from 'react'
import styled from 'styled-components'

interface ButtonProps {
  link?: boolean
  danger?: boolean
  onClick?: () => void
  primary?: boolean
  success?: boolean
  children: ReactNode
  block?: boolean
  active?: boolean
}

const ButtonStyle = styled.button<ButtonProps>`
  display: inline-flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  border: none;
  border-radius: 4px;
  padding: 4px 15px;
  font-size: 12px;
  text-decoration: none;
  background: #ffffff;
  border: 1px solid #e0e1e1;
  text-transform: uppercase;
  letter-spacing: 1px;
  ${(props) => props.block && `width: 100%;`}
  ${(props) =>
    props.primary &&
    `
    background: #469fe5;
    color: white;
  `}
  ${(props) =>
    props.success &&
    `
    background: #6bbb69;
    color: white;
  `}
  ${(props) =>
    props.danger &&
    `
    background: #EC476E;
    color: white;
  `}
  ${(props) =>
    props.active &&
    `
    background: #F2F3F4;
    color: #4F555C;
  `}
  &:hover {
    cursor: pointer;
    text-decoration: none;
  }
`

const ButtonLinkStyle = styled.button<ButtonProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  background: none;
  border: none;
  padding: 0;
  margin: 0;
  ${(props) => props.primary && `color: #469fe5;`}
  ${(props) => props.success && `color: #6bbb69;`}
  ${(props) => props.danger && `color: #EC476E;`}
  &:hover {
    cursor: pointer;
  }
`

const ButtonsGroupStyle = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
  > button {
    margin: 0 !important;
    border-radius: 0;
    &:first-child {
      border-radius: 4px 0 0 4px;
    }
    &:last-child {
      border-radius: 0 4px 4px 0;
    }
    &:not(:first-child) {
      border-left: 0;
    }
  }
`

export const Button: FC<ButtonProps> = (props) => {
  return props.link ? (
    <ButtonLinkStyle {...props}>{props.children}</ButtonLinkStyle>
  ) : (
    <ButtonStyle {...props}>{props.children}</ButtonStyle>
  )
}

export const ButtonsGroup: FC<{ children: ReactNode }> = ({ children }) => (
  <ButtonsGroupStyle>{children}</ButtonsGroupStyle>
)
