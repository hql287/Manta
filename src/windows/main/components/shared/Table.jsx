// Libraries
import React from 'react';

// Styles
import styled from 'styled-components';

const TableStyle = styled.table`
  width: 100%;
  ${props =>
    props.hasBorders &&
    `
    tr {
      border-bottom: 1px solid rgba(0,0,0,.1);
      &:last-child {
        border-bottom: none;
      }
    }
  `};
  ${props => props.sm && `tr { height: 36px }`};
  ${props => props.md && `tr { height: 42px }`};
  ${props => props.bg && `tr { height: 56px }`};
`;

const THeadStyle = styled.thead`
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  display: flex;
  justify-content: space-between;
  padding: 0 20px;
`;

const TBodyStyle = styled.tbody`
  display: flex;
  flex-direction: column;
  tr {
    padding-left: 20px;
    padding-right: 20px;
  }
`;

const THStyle = styled.th`
  text-transform: uppercase;
  font-size: 12px;
  color: #4f555c;
  letter-spacing: 1px;
  font-weight: 600;
  flex: 1;
  ${props =>
    props.actions &&
    `
    flex: none;
    width: 60px;
    justify-content: space-between;
  `};
`;

const TRStyle = styled.tr`
  display: flex;
  width: 100%;
  height: 40px;
  justify-content: space-between;
  align-items: center;
`;

const TDStyle = styled.td`
  font-size: 14px;
  flex: 1;
  color: #4f555c;
  height: 100%;
  display: flex;
  align-items: center;
  ${props => props.bold && `font-weight: 600;`};
  ${props => props.success && `color: #6BBB69;`};
  ${props => props.primary && `color: #469FE5;`};
  ${props => props.warning && `color: #F9D548;`};
  ${props => props.danger && `color: #EC476E;`};
  ${props => props.muted && `color: #B4B7BA;`};
  ${props =>
    props.actions &&
    `
    flex: none;
    width: 60px;
    justify-content: space-between;
  `};
`;

// Components
const Table = props => <TableStyle {...props}>{props.children}</TableStyle>;

const THead = props => <THeadStyle {...props}>{props.children}</THeadStyle>;

const TBody = props => <TBodyStyle {...props}>{props.children}</TBodyStyle>;

const TH = props => <THStyle {...props}>{props.children}</THStyle>;

const TR = props => <TRStyle {...props}>{props.children}</TRStyle>;

const TD = props => <TDStyle {...props}>{props.children}</TDStyle>;

export { Table, THead, TBody, TH, TR, TD };
