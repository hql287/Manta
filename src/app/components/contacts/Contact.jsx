// Libs
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

// Custom Components
import { TR, TD } from '../shared/Table';
import Button from '../shared/Button';

// Component
class Contact extends PureComponent {
  constructor(props) {
    super(props);
    this.deleteContact = this.deleteContact.bind(this);
    this.newInvoice = this.newInvoice.bind(this);
  }

  newInvoice() {
    const { newInvoice, contact } = this.props;
    newInvoice(contact);
  }

  deleteContact() {
    const { contact, deleteContact } = this.props;
    deleteContact(contact._id);
  }

  render() {
    const { contact } = this.props;
    return (
      <TR>
        <TD bold>{contact.fullname}</TD>
        <TD>{contact.email}</TD>
        <TD>{contact.phone}</TD>
        <TD actions>
          <Button link primary onClick={this.newInvoice}>
            <i className="ion-plus-round" />
          </Button>
          <Button link danger onClick={this.deleteContact}>
            <i className="ion-close-circled" />
          </Button>
        </TD>
      </TR>
    );
  }
}

Contact.propTypes = {
  contact: PropTypes.object.isRequired,
  deleteContact: PropTypes.func.isRequired,
  newInvoice: PropTypes.func.isRequired,
};

export default Contact;
