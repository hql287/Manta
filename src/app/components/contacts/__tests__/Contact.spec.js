// Libs
import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

// Component
import Contact from '../Contact';
import { TR, TD } from '../../shared/Table';
import Button from '../../shared/Button';

// Mocks
const contact = {
  _id: 'id-string',
  fullname: 'Jon Snow',
  email: 'jon@snow.hbo',
  phone: '000-000-000',
};
const newInvoice = jest.fn();
const deleteContact = jest.fn();

describe('Renders correctly to the DOM', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Contact
        contact={contact}
        newInvoice={newInvoice}
        deleteContact={deleteContact}
      />
    );
  });

  it('renders 1 TR row', () => {
    expect(wrapper.find(TR)).toHaveLength(1);
    expect(wrapper.find(TR)).not.toHaveLength(2);
  });

  it('renders 4 TD columns', () => {
    expect(wrapper.find(TD)).toHaveLength(4);
    expect(wrapper.find(TD)).not.toHaveLength(5);
  });

  it('receive correct props', () => {
    const mountWrapper = mount(
      <Contact
        contact={contact}
        newInvoice={newInvoice}
        deleteContact={deleteContact}
      />
    );
    expect(mountWrapper.prop('contact')).toEqual(contact);
    expect(mountWrapper.prop('newInvoice')).toEqual(newInvoice);
    expect(mountWrapper.prop('deleteContact')).toEqual(deleteContact);
  });

  it('renders 2 buttons in the last TD', () => {
    expect(
      wrapper
        .find(TD)
        .last()
        .find(Button)
    ).toHaveLength(2);
    expect(
      wrapper
        .find(TD)
        .last()
        .find(Button)
    ).not.toHaveLength(3);
  });

  describe('renders buttons correctly', () => {
    let firstBtn, lastBtn;
    beforeEach(() => {
      firstBtn = wrapper.find(Button).first();
      lastBtn = wrapper.find(Button).last();
    });

    it('renders buttons with correct props', () => {
      expect(firstBtn.prop('link')).toBe(true);
      expect(firstBtn.prop('primary')).toBe(true);
      expect(firstBtn.prop('success')).toBe(false);
      expect(lastBtn.prop('link')).toBe(true);
      expect(lastBtn.prop('danger')).toBe(true);
      expect(lastBtn.prop('success')).toBe(false);
    });

    it('handle click correctly', () => {
      firstBtn.simulate('click');
      expect(newInvoice).toHaveBeenCalled();
      expect(newInvoice).toHaveBeenCalledWith(contact);

      lastBtn.simulate('click');
      expect(deleteContact).toHaveBeenCalled();
      expect(deleteContact).toHaveBeenCalledWith('id-string');
    });
  });

  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <Contact
          contact={contact}
          newInvoice={newInvoice}
          deleteContact={deleteContact}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
