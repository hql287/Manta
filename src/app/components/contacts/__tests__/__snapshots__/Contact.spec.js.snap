// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly to the DOM matches snapshot 1`] = `
<tr
  className="Table__TRStyle-gHmqLx BngTM"
>
  <td
    className="Table__TDStyle-cAmfHP hlnQVD"
  >
    Jon Snow
  </td>
  <td
    className="Table__TDStyle-cAmfHP jrRCak"
  >
    jon@snow.hbo
  </td>
  <td
    className="Table__TDStyle-cAmfHP jrRCak"
  >
    000-000-000
  </td>
  <td
    className="Table__TDStyle-cAmfHP evw"
  >
    <button
      className="Button__ButtonLinkStyle-caVwCA lhMtRH"
      onClick={[Function]}
    >
      <i
        className="ion-plus-round"
      />
    </button>
    <button
      className="Button__ButtonLinkStyle-caVwCA gvxXrI"
      onClick={[Function]}
    >
      <i
        className="ion-close-circled"
      />
    </button>
  </td>
</tr>
`;
